### PR TITLE
WIP: implements #150 interface vlan mapping

### PR DESCRIPTION
**Pull request by lampwins**
### Fixes: #150 

The dcim `Interface` model as three new fields. `mode`, `native_vlan` and `tagged_vlan_members`. Note that `tagged_vlan_members` is a ManyToMany relationship to ipam `VLAN`. This includes all associated migration, view, form, model, and api changes. Submitted as a WIP for review.

### CODE REVIEW

**Comment by bdlamprecht** [comment]
@lampwins Looking forward for when this can be merged!

Note, please take these comments lightly, I'm not in any way attacking the great work it appears you've accomplished.

While **_briefly_** browsing the code, I believe this includes setting an interface to "access" or "trunk" or "trunk all".  Is "trunk all" the same as (in Cisco terms) `switchport mode trunk all`?

Not being very familiar with other vendors, I'm wondering if that `type` is necessary?
However, it is highly likely that I missed something in the code that requires that type to be there, which if is the case, please ignore my comments entirely and please accept my appreciation.

**Comment by collin-clark** [comment]
Trunk means different things to different manufacturers. In Cisco it means multiple VLAN's on an interface and in HP for example, a trunk is link aggregation.

**Comment by lampwins** [comment]
@bdlamprecht So the `Trunk All` mode is intended to mean all vlans are tagged. So yes, this would be cisco's `switchport mode trunk all` or `set interfaces ge-0/0/0.0 family ethernet-switching vlan members all` on junos. Whereas the `Trunk` mode allows the user to select specific vlans that should be tagged on the interface. `Trunk all` prohibits the user from selecting any tagged vlans from the list. 

Remember that netbox is intended to model the concepts of networking, not any one vendor's implementation. That being said, I am open to renaming `Trunk` and `Trunk all` if there is a better name.

**Comment by bdlamprecht** [comment]
Yeah, I suppose that makes sense.

I'm just wondering why we would want to "prohibit the user from selecting any tagged vlans" at all? I haven't seen how this looks in the UI, but in my mind, we are adding unnecessary complexity when  allowing them to select **all the vlans** would accomplish the same thing.  Then, let the vendor translate that into however it wants in the configuration.

Perhaps if I could see how this looks in the UI, I might understand more where you're coming from.

Concerning this statement:

> I am open to renaming Trunk and Trunk all if there is a better name.

If another name could be chosen to avoid confusion (for example in `collin-clark`'s example above), I would be all for it, however, I can't seem to come up with one either.

Any other ideas for the name?

**Comment by lampwins** [comment]
It is meant to say that all vlans are tagged so don't bother select anything. Each tagged vlan on an interface is a record in the database so we save the from having to manually tag potentially 4000 vlans.

Whats more, when it does come time to turn this into configuration, how would you know to use `vlan members all` instead of explicitly tagging all vlans?

**Comment by bdlamprecht** [comment]
> Each tagged vlan on an interface is a record in the database so we save the from having to manually tag potentially 4000 vlans.

That is a good point I didn't even consider.
I'm sold on the current implementation (and you didn't even have to show me a screenshot of the UI :smile: )

File: netbox/dcim/api/serializers.py [hunk 1]
```diff
@@ -670,7 +683,38 @@ class Meta:
...
+            device = data.get('device')
+
+        # Validate VLANs belong to the device's site or global
+        # We have to do this here decause of the ManyToMany relationship
+        native_vlan = data.get('native_vlan')
+        if native_vlan:
+            if native_vlan.site != device.site and native_vlan.site is not None:
+                raise serializers.ValidationError("Native VLAN is invalid for the interface's device.")
+
+        tagged_vlan_memebrs = data.get('tagged_vlan_members')
```
Comment (afics) [inline netbox/dcim/api/serializers.py]: Typo

**Comment by nonsbe** [comment]
Nice addition in the first place! 

@bdlamprecht @lampwins 
>I am open to renaming Trunk and Trunk all if there is a better name.

Why don't you give the choice untagged vlan - tagged vlans - tagged all? I think thats the most generic way all network engineers will understand. 
By the way if you have a trunk (in Cisco terms) it's possible to set your native (untagged) VLAN as well. How can you specify the `switchport trunk native vlan 200` in here?

**Comment by lampwins** [comment]
@nonsbe I can agree to that. The interface modes have been renames to `Access`, `Tagged` and `Tagged All`. 

> By the way if you have a trunk (in Cisco terms) it's possible to set your native (untagged) VLAN as well. How can you specify the switchport trunk native vlan 200 in here?

Yes this is supported. You can set an `untagged_vlan` as well as a set of `tagged_vlans`.

**Pull request merged**